odoo/addons/base/wizard/base_partner_merge.py

```diff
@@ -261,10 +263,28 @@ def write_serializer(item):
                             values[column] += write_serializer(item[column])
                         else:
                             values[column] = write_serializer(item[column])
+            elif field.company_dependent and field.name in summable_fields:
+                field_id = self.env['ir.model.fields']._get('res.partner', field.name).id
+                company_ids = self.env['ir.property'].sudo().search([
+                            ('fields_id', '=', field_id),
+                            ('res_id', 'in', res_id),
+                            ('company_id', '!=', False)
+                                ]).mapped('company_id')
+                for company_id in company_ids:
+                    if company_id not in values_company_dependent:
+                        values_company_dependent[company_id] = dict()
+                    for item in itertools.chain(src_partners, [dst_partner]):
+                        if item.with_company(company_id)[column]:
+                            if column in values_company_dependent[company_id]:
+                                values_company_dependent[company_id][column] += write_serializer(item.with_company(company_id)[column])
+                            else:
+                                values_company_dependent[company_id][column] = write_serializer(item.with_company(company_id)[column])
```

Comment:
Using `ir.property`'s model details is a big mistake IMO.
I suggest iterating on _all_ companies, and use `sudo()` (it is reasonable in this case).
With the `defaultdict` above, it becomes
```suggestion
                for company in self['res.company'].sudo().search([]):
                    for partner in (src_partners + dst_partner).sudo().with_company(company):
                        values_by_company[company][column] += write_serializer(partner[column])
```